### PR TITLE
Add EndGatewayCreationEvent

### PR DESCRIPTION
**Pull request by Joo200**
I want to cancel the creation of end gateways on my server after a dragon fight.
Can we please implement an event for that?

### CODE REVIEW

**Comment by Joo200** [comment]
Let me add some javadocs description from that event and fix a missing obfhelper-usage in nms code.
I just have to understand how to edit one single patch.

**Comment by MiniDigger** [comment]
To edit a patch, do an interactive rebase to that commit and ammend the commit. 
see https://github.com/PaperMC/Paper/blob/master/CONTRIBUTING.md#modifying-patches

File: Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+-            int i = (Integer) this.e.remove(this.e.size() - 1);
++            // Paper start - implement EndGatewayCreationEvent
++            int i = (Integer) this.getSpawnableGateways().get(this.getSpawnableGateways().size() - 1);
+             int j = (int) (96.0D * Math.cos(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
+             int k = (int) (96.0D * Math.sin(2.0D * (-3.141592653589793D + 0.15707963267948966D * (double) i)));
+-
++            Location loc = MCUtil.toLocation(this.getWorldServer(), j, 75, k);
++            EndGatewayCreationEvent event = new EndGatewayCreationEvent(loc.getBlock());
++            Bukkit.getPluginManager().callEvent(event);
++            if (event.isCancelled()) {
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch]:
```suggestion
+            if (!event.callEvent()) {
```
and remove the line above and the import

Comment (MiniDigger) [reply Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch]: in general, fully qualified imports are preferred as they reduce the size of the diff, so I would do that for the event and location.

**Review by MiniDigger — CHANGES_REQUESTED** [review]
minor stuff plus the missing usage of the obfhelper

**Comment by Joo200** [comment]
Thanks for the help.
I tried yesterday to edit the patch, without success.
I'll try it again with changed imports to simplify the patch and for the javadoc stuff in the event.

File: Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch [hunk 2]
```diff
@@ -47,7 +47,7 @@ index 09eabf123..87e6b63e5 100644
 +            Location loc = MCUtil.toLocation(this.getWorldServer(), j, 75, k);
 +            EndGatewayCreationEvent event = new EndGatewayCreationEvent(loc.getBlock());
 +            Bukkit.getPluginManager().callEvent(event);
```
Comment (mikroskeem) [inline Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch]: Get rid of this (as you're doing `event.callEvent()` below)

File: Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch [hunk 3]
```diff
@@ -47,7 +47,7 @@ index 09eabf123..87e6b63e5 100644
...
-+            if (event.isCancelled()) {
++            if (!event.callEvent()) {
```
Comment (mikroskeem) [inline Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch]: Maybe use `if(!new com.destroystokyo.paper.event.block.EndGatewayCreationEvent(loc.getBlock()).callEvent())` instead to reduce diff to smallest possible?

**Review by mikroskeem — APPROVED** [review]

File: Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch (hunk 3 above)
Comment (aikar) [reply Spigot-Server-Patches/0424-Added-EndGatewayCreationEvent.patch]: You're essentially double firing the event here. standard Bukkit callevent and this callevent. Do as mikro suggested.

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by BillyGalbreath** [comment]
~~If you moved the event logic out of `m()` and into `a(EntityEnderDragon entityenderdragon)` a few lines above it (where it calls `m()`)~~ Actually, just adding the dragon to `m()` as a parameter is easier, then you can include the dragon that died that caused the gateway creation. This could be useful for plugins in many ways.

**Comment by Joo200** [comment]
I made the requested changes.

**Comment by BillyGalbreath** [comment]
I'm beginning to wonder if you should use the currently unused EntityCreatePortalEvent and add a new PortalType.GATEWAY

I plan on adding back the event for the ENDER portal in the center of the island that md_5 removed in 1.9 for no apparent reason, and then remembered this PR. I think it fits.

**Comment by Joo200** [comment]
Feel free to refactor the EntityCreatePortalEvent. I don't have the time for change this PR the next weeks and I don't need the event right now.

You're right, the new event is not the best solution, maybe some sort of "StructureCreateEvent" could be better for structures like this.

**Comment by Joo200** [comment]
Closed.

The solution from @BillyGalbreath is better: https://github.com/PaperMC/Paper/pull/1986

**Pull request closed without merging**